README.md
# Teemo

Teemo is a dart library which provides an intuitive interface to the League of Legends LCU API

![](Teemo_demo.gif)

## Installing Teemo
Teemo is hosted on [pub.dev](pub.dev), which means you can get it with Flutter's built in `pub add` command.

```bash
$ flutter pub add teemo
```
This will add a line to your [pubspec.yaml](pubspec.yaml):
```
dependencies:
  teemo: ^0.2.4
```
**Note**: 0.2.4 was the current version at time of writing. It will show up in your pubspec as whatever is the most current version.

If you do not see this in your pubspec.yaml, run:
```bash
$ flutter pub get
```

Congratulations, you can now import Teemo into your Dart code with:
```dart
import 'package:teemo/teemo.dart';
```

## Using Teemo
Method documentation to help you with your development can be found [here](https://pub.dev/documentation/teemo/latest/teemo/teemo-library.html).


Since Teemo is asynchronous, you will need to use it inside asynchronous widgets like `FutureBuilder`. An example is below, and a working code sample can be found at [example/lib/main.dart](example/lib/main.dart).

```dart
FutureBuilder<Teemo>(
  future: _teemo, // a previously-obtained Future<String> or null
  builder: (BuildContext context, AsyncSnapshot<Teemo> snapshot) {
    List<Widget> children;
    if (snapshot.hasData) {
      children = <Widget>[
        OutlinedButton(
          onPressed: () {
            snapshot.data?.setCurrentRunePage(
                Rune.Conqueror,
                Rune.Triumph,
                Rune.LegendTenacity,
                Rune.LastStand,
                Rune.Transcendence,
                Rune.GatheringStorm,
                Rune.AdaptiveForcePerk,
                Rune.AdaptiveForcePerk,
                Rune.HealthPerk,
                name: 'Riven');
          },
          child: Text("Send Me To A Custom Lobby"),
        )
      ];
    } else {
      children = const <Widget>[
        SizedBox(
          width: 60,
          height: 60,
          child: CircularProgressIndicator(),
        ),
        Padding(
          padding: EdgeInsets.only(top: 16),
          child: Text('Awaiting result...'),
        )
      ];
    }
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: children,
      ),
    );
  },
)
```

## Understanding Events and Subscriptions
### Q: What are events?
### A:
Events are server side concepts. There are a many events you can subscribe to.
They are names for many different api updates that fall under their umbrella.
Use - `request('get', '/help')` - to get a JSON blob with a list of all possible events to subscribe to.
When events are triggered, they send the new data to Teemo.
Each update is of an endpoint underneath Event umbrella.
TL;DR
  Events are names for groups of endpoints.
  when an endpoint in an event changes, you get sent the new data for that endpoint.

### Q: How do I use events?
### A:
Subscribe to an event to begin receiving its messages. The `default_handler` argument
runs every time a message is received and not otherwise handled. You don't need to
supply a `default_handler`. If you don't the automatic behavior is to log it as info.

### Q: How do I interact with messages coming from events?
### A:
Messages come from endpoints. To catch an endpoint for special processing,
use `subscription_filter_endpoint` from Teemo, or `filter_endpoint` from the subscription.
these methods make the 'handler' argument run instead of the event's default handler.

###	REMEMBER TO MAKE EVENT HANDLER METHODS ASYNC

## Using the Websocket
Using the websocket is as simple as subscribing to an event and possibly filtering endpoints. Due to the callback nature of this architecture, all websocket actions are automatically called when the LCU sends a message that should be handled by a subscription or subscription filter.

## Making RESTful requests
Making rest requests is as simple as:
```Dart
Teemo teemo = await Teemo.create();
await teemo.request('POST', '/lol-lobby/v2/lobby', body: {
      "customGameLobby": {
        "configuration": {
          "gameMode": "PRACTICETOOL",
          "gameMutator": "",
          "gameServerRegion": "",
          "mapId": 11,
          "mutators": {"id": 1},
          "spectatorPolicy": "AllAllowed",
          "teamSize": 5
        },
        "lobbyName":"Name",
        "lobbyPassword":null
      },
      "isCustom":true
    });
```
This code snippet will send the user to a custom practice tool lobby.

### Fixing sandboxing issues on MacOS
Getting the error: "ps aux not authorized"? This is because by default MacOS apps are sandboxed and can't read the running processes. In your app's `macos/Runner/DebugProfile.entitlements` and `macos/Runner/Release.entitlements` comment out the line:
```
<!-- <key>com.apple.security.app-sandbox</key>
<true/> -->
```

---
*Teemo isn't endorsed by Riot Games and doesn't reflect the views or opinions of Riot Games or anyone officially involved in producing or managing Riot Games properties. Riot Games, and all associated properties are trademarks or registered trademarks of Riot Games, Inc.*


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.4] - 2022-03-23
### Fixed
- willump references

## [0.2.3] - 2022-03-23
### Fixed
- broken link

## [0.2.2] - 2022-03-23
### Added
- a proper README

## [0.2.1] - 2022-03-23
### Added
- added more docs
- added example program to set a runepage

## [0.2.0] - 2022-03-23
### Added
- complete method documentation for Teemo
- Runes class and ability for Teemo to set runepages

## [0.1.2] - 2022-03-03
### Fixed
- formatting with `flutter format .`

### Added
- post, put, head, update requests
- timeouts and retries to connecting to the LCU client if the app starts and there is not one to immediately connect to.

## [0.1.1] - 2022-03-03
### Fixed
- fixed linting in Teemo to bring naming in line with dart naming schemes

## [0.1.0] - 2022-03-03
### Added
- This Changelog
- Functional implementations of all Willump interfaces into Teemo


lib/src/runes.dart
/// Rune class, with static variables so one can reference runes by name.
class Rune {
  //Precision
  static const int _Precision = 8000;

  static const int PressTheAttack = 8005;
  static const int LethalTempo = 8008;
  static const int FleetFootwork = 8021;
  static const int Conqueror = 8010;

  static const int Overheal = 9101;
  static const int Triumph = 9111;
  static const int PresenceOfMind = 8009;

  static const int LegendAlacrity = 9104;
  static const int LegendTenacity = 9105;
  static const int LegendBloodline = 9103;

  static const int CoupDeGrace = 8014;
  static const int CutDown = 8017;
  static const int LastStand = 8299;

  //Domination
  static const int _Domination = 8100;

  static const int Electrocute = 8112;
  static const int Predator = 8124;
  static const int DarkHarvest = 8128;
  static const int HailOfBlades = 9923;

  static const int CheapShot = 8126;
  static const int TasteOfBlood = 8139;
  static const int SuddenImpact = 8143;

  static const int ZombieWard = 8136;
  static const int GhostPoro = 8120;
  static const int EyeballCollection = 8138;

  static const int RavenousHunter = 8135;
  static const int IngeniousHunter = 8134;
  static const int RelentlessHunter = 8105;
  static const int UltimateHunter = 8106;

  //Sorcery
  static const int _Sorcery = 8200;

  static const int SummonAery = 8214;
  static const int ArcaneComet = 8229;
  static const int PhaseRush = 8230;

  static const int NullifyingOrb = 8224;
  static const int ManaflowBand = 8226;
  static const int NimbusCloak = 8275;

  static const int Transcendence = 8210;
  static const int Celerity = 8234;
  static const int AbsoluteFocus = 8233;

  static const int Scorch = 8237;
  static const int Waterwalking = 8232;
  static const int GatheringStorm = 8236;

  //Resolve
  static const int _Resolve = 8400;

  static const int GraspOfTheUndying = 8437;
  static const int Aftershock = 8439;
  static const int Guardian = 8465;

  static const int Demolish = 8446;
  static const int FontOfLife = 8463;
  static const int ShieldBash = 8401;

  static const int Conditioning = 8429;
  static const int SecondWind = 8444;
  static const int BonePlating = 8473;

  static const int Overgrowth = 8451;
  static const int Revitalize = 8453;
  static const int Unflinching = 8242;

  //Inspiration
  static const int _Inspiration = 8300;

  static const int GlacialAugment = 8351;
  static const int UnsealedSpellbook = 8360;
  static const int FirstStrike = 8369;

  static const int HextechFlashtraption = 8306;
  static const int MagicalFootwear = 8304;
  static const int PerfectTiming = 8313;

  static const int FuturesMarket = 8321;
  static const int MinionDematerializer = 8316;
  static const int BiscuitDelivery = 8345;

  static const int CosmicInsight = 8347;
  static const int ApproachVelocity = 8410;
  static const int TimeWarpTonic = 8352;

  static const List<int> _PrecisionList = [
    PressTheAttack,
    LethalTempo,
    FleetFootwork,
    Conqueror,
    Overheal,
    Triumph,
    PresenceOfMind,
    LegendAlacrity,
    LegendTenacity,
    LegendBloodline,
    CoupDeGrace,
    CutDown,
    LastStand
  ];

  static const List<int> _DominationList = [
    Electrocute,
    Predator,
    DarkHarvest,
    HailOfBlades,
    CheapShot,
    TasteOfBlood,
    SuddenImpact,
    ZombieWard,
    GhostPoro,
    EyeballCollection,
    RavenousHunter,
    IngeniousHunter,
    RelentlessHunter,
    UltimateHunter
  ];

  static const List<int> _SorceryList = [
    SummonAery,
    ArcaneComet,
    PhaseRush,
    NullifyingOrb,
    ManaflowBand,
    NimbusCloak,
    Transcendence,
    Celerity,
    AbsoluteFocus,
    Scorch,
    Waterwalking,
    GatheringStorm
  ];

  static const List<int> _ResolveList = [
    GraspOfTheUndying,
    Aftershock,
    Guardian,
    Demolish,
    FontOfLife,
    ShieldBash,
    Conditioning,
    SecondWind,
    BonePlating,
    Overgrowth,
    Revitalize,
    Unflinching
  ];

  static const List<int> _InspirationList = [
    GlacialAugment,
    UnsealedSpellbook,
    FirstStrike,
    HextechFlashtraption,
    MagicalFootwear,
    PerfectTiming,
    FuturesMarket,
    MinionDematerializer,
    BiscuitDelivery,
    CosmicInsight,
    ApproachVelocity,
    TimeWarpTonic
  ];

  static const List<int> _Keystones = [
    PressTheAttack,
    LethalTempo,
    FleetFootwork,
    Conqueror,
    Electrocute,
    Predator,
    DarkHarvest,
    HailOfBlades,
    SummonAery,
    ArcaneComet,
    PhaseRush,
    GraspOfTheUndying,
    Aftershock,
    Guardian,
    GlacialAugment,
    UnsealedSpellbook,
    FirstStrike
  ];

  static const List<int> _MinorTier1 = [
    Overheal,
    Triumph,
    PresenceOfMind,
    CheapShot,
    TasteOfBlood,
    SuddenImpact,
    NullifyingOrb,
    ManaflowBand,
    NimbusCloak,
    Demolish,
    FontOfLife,
    ShieldBash,
    HextechFlashtraption,
    MagicalFootwear,
    PerfectTiming
  ];
  static const List<int> _MinorTier2 = [
    LegendAlacrity,
    LegendTenacity,
    LegendBloodline,
    ZombieWard,
    GhostPoro,
    EyeballCollection,
    Transcendence,
    Celerity,
    AbsoluteFocus,
    Conditioning,
    SecondWind,
    BonePlating,
    FuturesMarket,
    MinionDematerializer,
    BiscuitDelivery
  ];
  static const List<int> _MinorTier3 = [
    CoupDeGrace,
    CutDown,
    LastStand,
    RavenousHunter,
    IngeniousHunter,
    RelentlessHunter,
    UltimateHunter,
    Scorch,
    Waterwalking,
    GatheringStorm,
    Overgrowth,
    Revitalize,
    Unflinching,
    CosmicInsight,
    ApproachVelocity,
    TimeWarpTonic
  ];

  static const int HealthPerk = 5001;
  static const int AbilityHastePerk = 5007;
  static const int AttackSpeedPerk = 5005;
  static const int MagicResistPerk = 5003;
  static const int ArmorPerk = 5002;
  static const int AdaptiveForcePerk = 5008;

  static const List<List<int>> _PerkTiers = [
    [AdaptiveForcePerk, AttackSpeedPerk, AbilityHastePerk],
    [AdaptiveForcePerk, ArmorPerk, MagicResistPerk],
    [HealthPerk, ArmorPerk, MagicResistPerk]
  ];

  /// Validates a candidate page layout to make sure it's an allowed configuration.
  static bool validate(int keystone, int primary1, int primary2, int primary3,
      int secondary1, int secondary2, int perk1, int perk2, int perk3) {
    int primaryTree = Rune.treeId(keystone);
    int secondaryTree = Rune.treeId(secondary1);
    if (Rune.treeId(primary1) != primaryTree ||
        Rune.treeId(primary2) != primaryTree ||
        Rune.treeId(primary3) != primaryTree) return false;

    if (Rune.treeId(secondary2) != secondaryTree) return false;

    if (primaryTree == secondaryTree) return false;

    if (Rune.runeTier(keystone) != 0 ||
        Rune.runeTier(primary1) != 1 ||
        Rune.runeTier(primary2) != 2 ||
        Rune.runeTier(primary3) != 3) return false;

    if (Rune.runeTier(secondary1) >= Rune.runeTier(secondary2)) return false;

    if (!(Rune.perkTierValidate(perk1, 0) &&
        Rune.perkTierValidate(perk2, 1) &&
        Rune.perkTierValidate(perk3, 2))) return false;

    return true;
  }

  /// Gets the tier of a rune. Keystone: 0, rows of minor runes 1-3. Returns -1 on error.
  static int runeTier(int rune) {
    if (Rune._Keystones.contains(rune)) return 0;
    if (Rune._MinorTier1.contains(rune)) return 1;
    if (Rune._MinorTier2.contains(rune)) return 2;
    if (Rune._MinorTier3.contains(rune))
      return 3;
    else
      return -1;
  }

  /// Validates the minor perks. True on valid perk for the row, false on invalid.
  static bool perkTierValidate(int perk, int tier) {
    if (Rune._PerkTiers[tier].contains(perk))
      return true;
    else
      return false;
  }

  /// Returns the numeric ID of the tree for a given rune. Eg Taste Of Blood -> id of Domination tree.
  static int treeId(int rune) {
    if (Rune._PrecisionList.contains(rune))
      return Rune._Precision;
    else if (Rune._DominationList.contains(rune))
      return Rune._Domination;
    else if (Rune._SorceryList.contains(rune))
      return Rune._Sorcery;
    else if (Rune._ResolveList.contains(rune))
      return Rune._Resolve;
    else if (Rune._InspirationList.contains(rune))
      return Rune._Inspiration;
    else
      return -1;
  }
}


lib/src/teemo_impl.dart
library teemo;

import 'dart:io';
import 'dart:convert';
import 'package:flutter/services.dart' show rootBundle;

import 'package:teemo/src/EventSubscription.dart';
import 'package:teemo/src/runes.dart';

/// Teemo, the lcu interface. Use Teemo methods over [EventSubscription] methods or [Rune] methods.
class Teemo {
  String _authKey;
  int _port;
  HttpClient _restClient;
  WebSocket _websocket;

  Map<String, List<EventSubscription>> _subscriptions;

  Teemo._create(this._authKey, this._port, this._restClient, this._websocket)
      : _subscriptions = {};

  /// Asychronous constructor for Teemo. Attempts to connect to the LCU [retries] times, at [retryAfter] interval.
  ///
  /// Teemo will attempt to connect to the LCU websocket, as well as creating an HTTPS rest client instance.
  /// Teemo currently only uses the command-line arguments, not the lockfile
  static Future<Teemo> create({int retries = 5, int retryAfter = 5}) async {
    String authKey = '';
    int port = -1;

    for (int retry = 0;
        (authKey == '' || port == -1) && (retry < retries || retries == -1);
        retry++) {
      if (Platform.isMacOS || Platform.isLinux) {
        //we can run ps aux
        ProcessResult psRes = await Process.run('ps', ['aux']);
        List<String> args = psRes.stdout.split(' ');

        for (int i = 0; i < args.length; i++) {
          List<String> keyVal = args[i].split('=');
          if (keyVal[0] == '--remoting-auth-token') {
            authKey = keyVal[1];
          } else if (keyVal[0] == '--app-port') {
            try {
              port = int.parse(keyVal[1]);
            } catch (error) {
              print('couldn\'t parse port to number');
              break;
            }
          }
          if (authKey != '' && port >= 0) break;
        }
      }
      if (authKey != '' && port >= 0) break;
      /* print('waiting, will retry'); */
      await new Future.delayed(Duration(seconds: retryAfter));
      /* print('retrying: (${retries - retry} retries remaining)'); */
    }

    if (authKey == '' || port == -1) print('something\'s gone wrong');

    print(authKey + ':' + port.toString());

    String cert =
        await rootBundle.loadString('packages/teemo/assets/riotgames.pem');
    print(cert);
    SecurityContext secCtx = SecurityContext();
    secCtx.setTrustedCertificatesBytes(utf8.encode(cert));

    Teemo teemo = Teemo._create(
        authKey,
        port,
        HttpClient(context: secCtx),
        await WebSocket.connect('wss://127.0.0.1:$port',
            headers: {
              'Authorization':
                  'Basic ' + utf8.fuse(base64).encode('riot:$authKey')
            },
            customClient:
                HttpClient(context: secCtx) //TODO: try teemo.rest_client
            ));
    teemo._startWebsocketListener();
    return teemo;
  }

  void _startWebsocketListener() {
    this._websocket.listen((data) {
      if (data.isEmpty) return;
      List<dynamic> jsonData = json.decode(data);
      List<EventSubscription> subscriptions =
          this._subscriptions[jsonData[1]] ?? [];
      subscriptions.forEach((sub) => sub.tasks(jsonData[2]));
    });
  }

  /// Sends a subscription request for [event] to the LCU. When that event is recieved in the future [subscription] is launched.
  ///
  /// [behavior] and [subscription] are optional parameters for complex logic flows where you want to send multiple events through one subscription
  /// [behavior] is a shortcut to passing itself to the subscription constructor.
  /// [subscription], if null, is created and returned from the function. Else, it is returned from the function as is.
  EventSubscription subscribe(String event,
      {EventSubscription? subscription, Handler? behavior}) {
    EventSubscription _subscription = subscription ??
        (behavior != null ? EventSubscription(behavior) : EventSubscription());
    this._subscriptions.update(event, (list) => list + [_subscription],
        ifAbsent: () {
      this._websocket.add('[${EventCode.SUBSCRIBE}, "$event"]');
      return [_subscription];
    });
    return _subscription;
  }

  /// Unsubscribes [event]. If [subscription] is provided, only that subscription is removed from the event. Otherwise ALL subscriptions are removed and the event is unsubscribed at the LCU level.
  /// If the last subscription is removed from an event, it is unsubscribed at the LCU level.
  /// It is possible to unsubscribe to events that were never subscribed to. This is ignored by the LCU.
  void unsubscribe(String event, {EventSubscription? subscription}) {
    if (subscription != null) {
      this._subscriptions[event]?.remove(subscription);
      if (!(this._subscriptions[event]?.isEmpty ?? true)) {
        //WEBSOCKET DOESN'T ADD UNSUBSCRIBE IN THIS BRANCH
        return;
      }
    }
    this._subscriptions.remove(event);
    this._websocket.add('[${EventCode.UNSUBSCRIBE}, $event]');
  }

  /// Returns the [List<EventSubscription>] from Teemo's database of all subscriptions attached to [event].
  List<EventSubscription> getEventSubscription(String event) {
    return this._subscriptions[event] ?? [];
  }

  /// Adds a special callback to [subscription] which triggers when an event with the uri [endpoint] gets sent from the LCU.
  /// If no [behavior] is provided, the subscription's default handler is used. (This is effectively the same as never calling this function, however it is valid.)
  void subscriptionFilterEndpoint(
      EventSubscription subscription, String endpoint,
      {Handler? behavior}) {
    behavior != null
        ? subscription.filterEndpoint(endpoint, behavior: behavior)
        : subscription.filterEndpoint(endpoint);
  }

  /// Removes the special behavior endpoint from the subscription. If absent, while valid, nothing happens.
  void subscriptionUnfilterEndpoint(
      EventSubscription subscription, String endpoint) {
    subscription.unfilterEndpoint(endpoint);
  }

  /// Used to send GET, POST, HEAD, etc... requests.
  Future<JsonData> request(String method, String endpoint,
      {JsonData? body}) async {
    HttpClientRequest req = await this
        ._restClient
        .openUrl(method, Uri.parse('https://127.0.0.1:${this._port}$endpoint'));

    req.headers.set(HttpHeaders.acceptHeader, 'application/json');
    req.headers.set(HttpHeaders.contentTypeHeader, 'application/json');
    req.headers.set(HttpHeaders.authorizationHeader,
        'Basic ' + utf8.fuse(base64).encode('riot:${this._authKey}'));

    if (body != null) req.add(json.fuse(utf8).encode(body));

    HttpClientResponse resp = await req.close();
    return resp
        .transform(utf8.decoder)
        .join()
        .then((String data) => json.decode(data.isEmpty ? '{}' : data));
  }

  /// Attempts to close Teemo's websocket and rest client. Call before exiting the app.
  void close() {
    this._websocket.close();
    this._restClient.close();
  }

  /// Sets current rune page using either ids or [Rune] static members. If an invalid page is requested, nothing happens.
  void setCurrentRunePage(
      int keystone,
      int primary1,
      int primary2,
      int primary3,
      int secondary1,
      int secondary2,
      int perk1,
      int perk2,
      int perk3,
      {String name = 'Teemo Created Page'}) async {
    if (!Rune.validate(keystone, primary1, primary2, primary3, secondary1,
        secondary2, perk1, perk2, perk3)) return;

    JsonData currentPage =
        await this.request('GET', '/lol-perks/v1/currentpage');
    /* print(currentPage); */
    currentPage = await this
        .request('DELETE', '/lol-perks/v1/pages/${currentPage['id']}');
    /* print(currentPage); */
    currentPage = await this.request('POST', '/lol-perks/v1/pages', body: {
      "name": name,
      "primaryStyleId": Rune.treeId(keystone),
      "subStyleId": Rune.treeId(secondary1),
      "selectedPerkIds": [
        keystone,
        primary1,
        primary2,
        primary3,
        secondary1,
        secondary2,
        perk1,
        perk2,
        perk3
      ],
      "current": true
    });
    /* print(currentPage); */
  }
}


lib/src/EventSubscription.dart
typedef JsonData = Map<String, dynamic>;

/// The type of ALL Teemo behavior callbacks. an asynchronous void function which consumes JSON.
typedef Handler = Future<void> Function(JsonData data);

/// The three LCU ids for requesting levels of activity on events.
class EventCode {
  static const int SUBSCRIBE = 5;
  static const int UNSUBSCRIBE = 6;
  static const int RESPONSE = 8;
}

/// EventSubscription methods can be used to manipulate subscriptions, however Teemo provides all the necessary abstraction.
class EventSubscription {
  Map<String, List<Handler>> _registeredUris = {};
  Map<String, List<Handler>> _registeredPaths = {};
  Handler _defaultBehavior;

  static Future<void> _defaultBehaviorImpl(JsonData data) async {
    print(data);
  }

  /// Default constructor uses the private defaultbehavior handler for all unhandled endpoints, which is to print all JSON consumed.
  /// Provide a [Handler] argument to override the subscription's default behavior on all incoming messages.
  EventSubscription(
      [this._defaultBehavior = EventSubscription._defaultBehaviorImpl]);

  /// filters an endpoint to be consumed with special behavior. if [endpoint] ends in '*', all URIs starting with endpoint are consumed by this filter.
  void filterEndpoint(String endpoint,
      {Handler behavior = EventSubscription._defaultBehaviorImpl}) {
    if (endpoint.endsWith('*')) {
      //then this is a path and we want to glob match
      this._registeredPaths.update(endpoint.substring(0, endpoint.length - 1),
          (list) => list + [behavior],
          ifAbsent: () => [behavior]);
    } else {
      this._registeredUris.update(endpoint, (list) => list + [behavior],
          ifAbsent: () => [behavior]);
    }
  }

  /// unfilters an endpoint. To unfilter an endpoint which ended in '*', provide the '*' here too.
  void unfilterEndpoint(String endpoint) {
    if (endpoint.endsWith('*')) {
      //then this is a path and we want to glob match
      this._registeredPaths.remove(endpoint.substring(0, endpoint.length - 1));
    } else {
      this._registeredUris.remove(endpoint);
    }
  }

  /// A mostly private method for Teemo to gather all handlers which match the incoming [data] from the LCU server.
  List<Future<void>> tasks(JsonData data) {
    List<Handler> tasks = [];

    this._registeredPaths.forEach((key, value) =>
        {data['uri'].startsWith(key) ? tasks.addAll(value) : {}});

    tasks.addAll(this._registeredUris[data['uri']] ?? []);

    if (tasks.isEmpty) tasks.add(this._defaultBehavior);
    return tasks.map((e) => e(data)).toList();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:teemo/teemo.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyStatefulWidget(),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({Key? key}) : super(key: key);

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  final Future<Teemo> _teemo = Teemo.create();

  @override
  Widget build(BuildContext context) {
    return DefaultTextStyle(
      style: Theme.of(context).textTheme.headline2!,
      textAlign: TextAlign.center,
      child: FutureBuilder<Teemo>(
        future: _teemo, // a previously-obtained Future<String> or null
        builder: (BuildContext context, AsyncSnapshot<Teemo> snapshot) {
          List<Widget> children;
          if (snapshot.hasData) {
            children = <Widget>[
              OutlinedButton(
                onPressed: () {
                  snapshot.data?.setCurrentRunePage(
                      Rune.Conqueror,
                      Rune.Triumph,
                      Rune.LegendTenacity,
                      Rune.LastStand,
                      Rune.Transcendence,
                      Rune.GatheringStorm,
                      Rune.AdaptiveForcePerk,
                      Rune.AdaptiveForcePerk,
                      Rune.HealthPerk,
                      name: 'Riven');
                },
                child: Text("Send Me To A Custom Lobby"),
              )
            ];
          } else if (snapshot.hasError) {
            children = <Widget>[
              const Icon(
                Icons.error_outline,
                color: Colors.red,
                size: 60,
              ),
              Padding(
                padding: const EdgeInsets.only(top: 16),
                child: Text('Error: ${snapshot.error}'),
              )
            ];
          } else {
            children = const <Widget>[
              SizedBox(
                width: 60,
                height: 60,
                child: CircularProgressIndicator(),
              ),
              Padding(
                padding: EdgeInsets.only(top: 16),
                child: Text('Awaiting result...'),
              )
            ];
          }
          return Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: children,
            ),
          );
        },
      ),
    );
  }
}
